### DISCUSSION TITLE
How to persist and display data with Aspire standalone dashboard

### DISCUSSION THREAD

**Discussion by SirPustekuchen** [microsoft/aspire#2564] [Q&A] [answered]
Hello Aspire Team,

I have tried the new standalone dashboard and I think it's great! Great work and I'm excited to see what else is coming.

As I am not yet an expert in the .NET Core Framework myself, I have a question. In general, the issue is that currently displayed data is gone after restarting the dashboard container. How can I make it so that such data can be saved and then displayed again so that I can also see historical data? What is the recommended way when using the standalone dashboard?

If this could somehow work without using the whole stack from e.g. Prometheus, Grafana, Tempo, Loki, that would be great. I favour a lightweight solution.

Best regards and great work
Wolfram

**Comment by davidfowl** [comment, marked as answer]
There's no support for persisting the telemetry data and no plans to implement external storage. When you restart the dashboard, the data is lost. Treat this dashboard like your volatile telemetry source, not a long-term data store for telemetry.

For historical data, you need a real production ready telemetry system. The dashboard outside of developer scenarios works best as *one* of the targets for telemetry data instead of being the only one (especially if you want historical data).

**Reply by SirPustekuchen** [reply]
Many thanks for the quick reply :-)

**Reply by expcat** [reply]
We hope that future versions will consider adding a feature for storing telemetry data. This is particularly important for smaller-scale application projects, which, in the early stages of development, often do not have a large number of services and telemetry data. However, to achieve the storage and display of telemetry data, it is necessary to deploy a significant amount of middleware services, which can be a challenge for projects with limited resources. Therefore, a simplified telemetry monitoring service that supports data persistence would be greatly beneficial, especially for developers in the early stages of their projects.

**Reply by davidfowl** [reply]
I can't predict what will happen in the future, but we have no plans, even for small shops, to persist the data outside of the dashboard. Building a durable telemetry storage system is a project by itself, the UI is easy in comparison.

**Reply by b4ux1t3** [reply]
I don’t mean to bring up the past, but it’s been a year and a half now and I’m wondering if this is still the stance 

Aspire keeps getting billed as something I can use to run dotnet apps _in production_. 

Is the actual guidance here “make all your applications require additional configuration to support exporting telemetry even though we’re already aggregating them in one place”?

It seems like simply allowing aspire itself to integrate with existing tools shouldn’t be all that complex. In the end, I don’t actually care if Aspire is persisting the logs, I just care that they’re getting persisted, and Aspire is already aggregating them all; is there no world where it could just forward them along?

I hope this doesn’t come across as rude or anything, it’s just that this one feature would turn Aspire from “meh, it’s neat” to _the_ thing I would recommend for running dotnet-based infrastructure on.

**Reply by davidfowl** [reply]
See the roadmap https://github.com/dotnet/aspire/discussions/10644

> is there no world where it could just forward them along?

By this I am assuming you are looking for something free and convenient that work as well as the aspire dashboard, but also stores telemetry yes?

**Reply by b4ux1t3** [reply]
Oh, I had not seen that item on the roadmap! That's excellent news, at least for development.

I'm not sure I care about free or even easy, I just ran into the issue today and was diving down the rabbit hole of trying to work around it.

I'd _like_ something free and convenient, obviously, that I could just say "Hey, Aspire, point your hosted projects to this telemetry service". 

That said, like, I'd settle for Aspire publishing a prometheus endpoint that can be scraped and pull down the telemetry it's already collecting, you know?

Mostly I just wasn't sure about the state and was a little frustrated with the topic. Haha. Really great to see the progress you guys are making, thank you for engaging on a Sunday night!

**Reply by davidfowl** [reply]
> That said, like, I'd settle for Aspire publishing a prometheus endpoint that can be scraped and pull down the telemetry it's already collecting, you know?

One of the benfits of using a standard like otel is that a LOT of different vendors support it so you should be able to point your application at another agent or vendor that supports OTLP and your telemetry will go elsewhere. 

There's a new otel collector integration coming to the toolkit soon https://github.com/CommunityToolkit/Aspire/pull/603. If you are unfamilar https://opentelemetry.io/docs/collector/, which enables pushing telemetry to N different sinks.

Finally, here's a sample that shows how to manually wire up prometheus and grafana using the otel collector https://github.com/dotnet/aspire-samples/tree/main/samples/Metrics

**Reply by VladislavAntonyuk-AccuReference** [reply]
@davidfowl is there a way to add a feature to export logs? It will be easy for us to export the logs we need.

**Comment by kannan-cidc** [comment]
@expcat  @SirPustekuchen  - came across this article, hope this helps 

https://blog.datalust.co/persistent-logs-and-traces-for-the-net-aspire-dashboard/ 

- cheers
